### PR TITLE
CATROID-570 Refactor DeleteItemOfUserListActionTest to Kotlin

### PR DESCRIPTION
**Pull request by trummerp**
- Refactored DeleteItemOfUserListActionTest to Kotlin
- Added some tests
- Reworked some tests
- Fixed a bug in DeleteItemOfUserListAction, where invalid parameter caused a deletion of the first element

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+import java.util.ArrayList
+
+@RunWith(JUnit4::class)
+class DeleteItemOfUserListActionTest {
+    private var action: DeleteItemOfUserListAction? = null
+    private var userList: UserList? = null
+    private var collection: ArrayList<Any>? = null
```
Comment (ndrnour) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt:41]: please change `action, userList, collection` to `lateinit var` and remove not-null assertion operators `!!` from your code

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    private var initialCollectionSize = 0
+
+    @Before
+    @Throws(Exception::class)
+    fun setUp() {
+        action = DeleteItemOfUserListAction()
+        collection = ArrayList(listOf(1.0, "2.0", "test", 4.0))
```
Comment (ndrnour) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt:48]: change to `collection = arrayListOf(1.0, "2.0", "test", 4.0)`

File: catroid/src/main/java/org/catrobat/catroid/content/actions/DeleteItemOfUserListAction.kt [hunk 3]
```diff
@@ -38,10 +38,10 @@ class DeleteItemOfUserListAction : TemporalAction() {
         val listSize = userList?.value?.size.takeUnless { it == 0 } ?: return
 
         val indexToDelete = try {
-            formulaIndexToDelete?.interpretInteger(sprite)?.minus(1) ?: 0
+            formulaIndexToDelete?.interpretInteger(sprite)?.minus(1) ?: -1
```
Comment (ndrnour) [inline catroid/src/main/java/org/catrobat/catroid/content/actions/DeleteItemOfUserListAction.kt:41]: 👍🏼

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+@RunWith(JUnit4::class)
+class DeleteItemOfUserListActionTest {
+    private var action: DeleteItemOfUserListAction? = null
+    private var userList: UserList? = null
+    private var collection: ArrayList<Any>? = null
+    private var initialCollectionSize = 0
+
+    @Before
+    @Throws(Exception::class)
```
Comment (ndrnour) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt:45]: is this really needed? `@Throws(Exception::class)`

**Review by ndrnour — CHANGES_REQUESTED** [review]
last modification. 💯

**Review by Koell — APPROVED** [review]
Thanks

**Comment by wslany** [comment]
> last modification. 💯

Has this been resolved? If so, can you add your approval?

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    companion object {
+        @JvmStatic
+        @Parameterized.Parameters(name = "{0}")
+        fun parameters() = listOf(
+            arrayOf("DELETE_SINGLE", listOf(Formula(1)), 3),
+            arrayOf("DELETE_MULTIPLE", listOf(Formula(4), Formula(1)), 2),
+            arrayOf("DELETE_ALL", listOf(Formula(4), Formula(1), Formula(2), Formula(1)), 0),
+            arrayOf("NUMERICAL_STRING", listOf(Formula("1")), 3),
+            arrayOf("NUMBER_NEGATIVE", listOf(Formula(-1)), 4),
```
Comment (wslany) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt:53]:
```suggestion
            arrayOf("NUMBER_NEGATIVE", listOf(Formula(-1)), 4),
            arrayOf("NUMBER_TOO_LARGE", listOf(Formula(5)), 4),
```

Missing case.

**Review by wslany — CHANGES_REQUESTED** [review]
Thanks. However, the tests in Java were more comprehensive: they tested, e.g., also whether the _intended_ elements were deleted, whereas in the new tests in Kotlin, you test only if the number of elements is correct. It could be that the order of deletion uses the indices in reverse, or is somehow random, and these tests would still be green, whereas the old test suite in Java would have caught that.

Having said that, it is anyway necessary to add a CLT for this brick, and then the more complex tests can be relegated there. Please add such a CLT for the brick in https://github.com/Catrobat/Catroid/tree/develop/catroid/src/androidTest/assets/catrobatTests/data instead of adding them to the Kotlin test. And please also add the test cases from the Kotlin tests to the CLTs. 

We want to use the CLTs as a formal, executable, but device independent specification in the form of automatic tests of the Catrobat language. Catty and other projects will be able to execute them, and any future implementation can use them as a reference against which to implement it.

**Comment by trummerp** [comment]
Update: 
- Added missing case to the kotlin testcases
- Added a CLT including all possible kotlin testcases, as well as one for the deletion order

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import java.util.ArrayList
+
+@RunWith(Parameterized::class)
+class DeleteItemOfUserListActionTest(
+    private val name: String,
+    private val deleteList: List<Formula?>?,
```
Comment (gPathpp) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt:38]:
```suggestion
    private val deleteList: List<Formula?>,
```

File: catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    fun setUp() {
+        action = DeleteItemOfUserListAction()
+        collection = arrayListOf(1.0, "2.0", "test", 4.0)
+        userList = UserList("userList", collection)
+        action.userList = userList
+    }
+
+    @Test
+    fun testDeleteItemOfUserList() {
+        for (idx in deleteList!!) {
```
Comment (gPathpp) [inline catroid/src/test/java/org/catrobat/catroid/test/content/actions/DeleteItemOfUserListActionTest.kt:72]:
```suggestion
        for (idx in deleteList) {
```

**Review by gPathpp — CHANGES_REQUESTED** [review]
PLEASE-ADD: a JUNIT test that verifies that the right index was deleted.
Also, perform a rebase, please.

**Pull request closed without merging**